### PR TITLE
Establish pattern for exposing attributes in sysfs

### PR DESCRIPTION
**Pull request by joelsmithTT**
Expose card identification information and clock speeds.

Such information does not fit into the hwmon framework.

### CODE REVIEW

File: hwmon.c [hunk 1]
```diff
@@ -69,3 +70,65 @@ const struct hwmon_ops tt_hwmon_ops = {
...
+	u64 card_type = (board_id_hi >> 4) & 0xFFFF;
+	if (card_type == 0x3)
+		return scnprintf(buf, PAGE_SIZE, "e150\n");
+	if (card_type == 0x7)
+		return scnprintf(buf, PAGE_SIZE, "e75\n");
+	if (card_type == 0x14)
+		return scnprintf(buf, PAGE_SIZE, "n300\n");
+	if (card_type == 0x18)
+		return scnprintf(buf, PAGE_SIZE, "n150\n");
+	return scnprintf(buf, PAGE_SIZE, "unknown\n");
```
Comment (alewycky-tenstorrent) [inline hwmon.c:98]: Switch statement to pick string.

File: grayskull.c [hunk 2]
```diff
@@ -135,6 +135,29 @@ struct TLB_16M_REG {
 
 #define TLB_16M_SIZE_SHIFT 24
 
+
+static struct tt_attribute_data gs_attributes[] = {
```
Comment (alewycky-tenstorrent) [inline grayskull.c:139]: Can you just put a pointer into tenstorrent_device_class and move init/cleanup to device-indepdent code? Terminate the array with __ATTR_NULL.

File: hwmon.c [hunk 3]
```diff
@@ -69,3 +70,65 @@ const struct hwmon_ops tt_hwmon_ops = {
...
+	struct tenstorrent_device *tt_dev = dev_get_drvdata(dev);
+	struct tt_attribute_data *data = container_of(attr, struct tt_attribute_data, attr);
+	struct tt_hwmon_context *ctx = &tt_dev->hwmon_context;
+
+	u32 board_id_hi = ioread32(ctx->telemetry_base + data->reg_offset);
+	u32 board_id_lo = ioread32(ctx->telemetry_base + data->reg_offset + 4);
+	return scnprintf(buf, PAGE_SIZE, "%08X%08X\n", board_id_hi, board_id_lo);
+}
+
+ssize_t tt_show_arc_fw(struct device *dev, struct device_attribute *attr, char *buf) {
```
Comment (alewycky-tenstorrent) [inline hwmon.c:111]: Needs rename because it's used for other-than-arc fw.

File: enumerate.c
Comment (alewycky-tenstorrent) [inline enumerate.c:1]: Ideally break into separate commit so the commit message can explain the change.

Comment (joelsmithTT) [reply enumerate.c:1]: If you were referring to the reordering of some function calls in enumerate.c -- that turned out not to be needed.

File: hwmon.c (hunk 1 above)
Comment (alewycky-tenstorrent) [reply hwmon.c]: I was thinking more like
```
switch (card_type) {
    case 0x03: card_name = "e150"; break;
...
    default: card_name = "unknown"; break;
}

return scnprintf(buf, PAGE_SIZE, "%s\n", card_name);
```

File: enumerate.c [hunk 4]
```diff
@@ -96,13 +97,30 @@ static int tenstorrent_pci_probe(struct pci_dev *dev, const struct pci_device_id
 		register_reboot_notifier(&tt_dev->reboot_notifier);
 	}
 
+	if (tt_dev->attributes) {
+		struct tt_attribute_data *data = tt_dev->attributes;
+		while (data->attr.attr.name) {
+			device_create_file(&tt_dev->dev, &data->attr);
+			data++;
+		}
```
Comment (alewycky-tenstorrent) [inline enumerate.c:105]: I would prefer if the iteration is in a for loop like `for (; data->attr.attr.name; data++) {`. Ditto in tenstorrent_pci_remove.

File: hwmon.c (hunk 1 above)
Comment (joelsmithTT) [reply hwmon.c]: Fixed

File: hwmon.c [hunk 5]
```diff
@@ -69,3 +70,66 @@ const struct hwmon_ops tt_hwmon_ops = {
...
+	return sprintf(buf, "%u\n", value);
+}
+
+ssize_t tt_show_card_type(struct device *dev, struct device_attribute *attr, char *buf) {
+	struct tenstorrent_device *tt_dev = dev_get_drvdata(dev);
+	struct tt_attribute_data *data = container_of(attr, struct tt_attribute_data, attr);
+	struct tt_hwmon_context *ctx = &tt_dev->hwmon_context;
+
+	u32 board_id_hi = ioread32(ctx->telemetry_base + data->reg_offset);
+	u64 card_type = (board_id_hi >> 4) & 0xFFFF;
```
Comment (alewycky-tenstorrent) [inline hwmon.c:89]: Why is card_type u64? Seems it can't be more than 16 bits.

Comment (joelsmithTT) [reply hwmon.c]: Fixed

**Review by alewycky-tenstorrent — APPROVED** [review]
